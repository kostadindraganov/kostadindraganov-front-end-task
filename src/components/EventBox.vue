<template>
  <div 
  class="event-box absolute z-10 bg-white border-l-4 border-pink-500"
  :style="style"
  >
    <h3 class="title">
     Event title
    </h3>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';
import { Events } from '@/types/Events';

@Component
export default class EventBox extends Vue {
  @Prop({ 
    type: Object, 
    default: {} 
  }) private event!: Events;

get style() {
  return `top:${this.event.top}px; left:${this.event.left}px; width:${this.event.width}px; height:${this.event.height}px;`
}

}
</script>
