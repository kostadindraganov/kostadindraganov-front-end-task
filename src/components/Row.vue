<template>
  <div class="time-row relative w-100" ref="row">
    <div class="absolute -left-20 -top-33">
      <div class="w-16 text-right">
      <span class="text-lg text-slate-500"> {{title}} </span>

      </div>
    </div>

    <div v-if="showRow" class="time-body bg-gray-100 h-15.5 border-l-2 border-l-slate-200 box-border">

    </div>
  
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';

@Component
export default class Row extends Vue {
  @Prop() private title!: string;
  @Prop({ 
    type: Boolean, 
    default: false 
  }) private isLast!: boolean;

  get showRow(): boolean {
    return !this.isLast;
  }

}
</script>
