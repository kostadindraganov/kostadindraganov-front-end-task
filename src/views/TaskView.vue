<template>
  <main class="container flex flex-col items-center mb-10">
    <h1>Jiminny Calendar</h1>

    <section>
      <div class="relative mx-3">
        <ul>
          <li v-for="(event, index) in eventsBox" :key="index">
            <event-box :event="event" class="animate-fade-in-up animation-delay-200"/>
          </li>
        </ul> 
      </div>    
      <ul>
        <li v-for="(item, index) in allTimeFrames" :key="index">
         <row :title="item" :isLast="lastItem === index"/>
        </li>
      </ul>
    </section>
    
  </main>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import {Action, Getter} from 'vuex-class';
import moment from 'moment'

import Row from '@/components/Row.vue'; 
import EventBox from '@/components/EventBox.vue';
import timeFrame from '@/constants/timeFrame';

import { Event } from '@/types/Event';
import { Events } from '@/types/Events';

const namespace = 'events';

@Component({
  components: {
    Row,
    EventBox
  },
})
export default class TaskView extends Vue {
  @Action('fetchEvents', { namespace })
  fetchEvents!: () => Promise<Event[]>;
  /* eslint-disable @typescript-eslint/no-explicit-any */
  @Getter('getEvents', { namespace })
  getEvents!: any;

  width = 720;

  height = 840; 

  tableWidth = 696;

  endOfEvent = 0;

  eventsBox: Events[] = [];

  get allTimeFrames(): string[] {
    return timeFrame.map((item: number) => moment.unix(item).utcOffset('+0300').format('h A'));
  }

  get lastItem() {
    return this.allTimeFrames.length - 1;
  }

  created() {
    this.getAllEvents()
  }

 async getAllEvents(): Promise<void> {
    await this.fetchEvents();

    if (Array.isArray(this.getEvents) && this.getEvents.length > 0) {
      this.sortEvents();
      this.calcPosition();
    }
  }

  sortEvents(): void {
    const events = this.getEvents.sort((first: Event,last: Event) => {
      if(first.startTime > last.startTime) { return 1; }
      if(first.startTime < last.startTime) { return -1; }
      return 0;
    });

    const transformedEvents: Events[] = events.map((item: Event) => 
      Object.assign(item, {
        height: item.endTime - item.startTime,
        top: item.startTime
      })
    )

    this.eventsBox = transformedEvents;
  }

  checkOverlapItems(events):  Array<number[]> {
    let eventsOrder: Array<number[]> = [];
    this.endOfEvent = events[0].endTime;

    const childOrder = events.reduce((acc: number[] = [0], item: number, index: number) => {
      if(events[index].startTime <= this.endOfEvent){ 
        if(events[index].endTime >= this.endOfEvent) { 
          this.endOfEvent = events[index].endTime;
        } 
        acc.push(index); 
      } else {
        eventsOrder.push(acc);
        acc = []; 
        acc.push(index); 
        this.endOfEvent = events[index].endTime;
      }
      return acc;
    }, [])

    eventsOrder.push(childOrder);
    
    return eventsOrder;
  }

  calcPosition(): void {
    const eventsOrder: Array<number[]> = this.checkOverlapItems(this.getEvents);
    const eventsBox: Events[] = this.eventsBox;
    let index: number;

    if (eventsOrder.length) {
      for (let item of eventsOrder) {
        index = item.length;
        item.map((el, key) => {
          eventsBox[el].width = this.tableWidth/index;
          eventsBox[el].left = this.tableWidth/index*key;
        })
      }
    }

    this.eventsBox = eventsBox
  }
}
</script>
