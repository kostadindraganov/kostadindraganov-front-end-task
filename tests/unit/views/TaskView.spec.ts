import { shallowMount, createLocalVue, RouterLinkStub } from '@vue/test-utils'
import Vuex from 'vuex'
import TaskView from '@/views/TaskView.vue';
import Row from '@/components/Row.vue';
import EventBox from '@/components/EventBox.vue';

const localVue = createLocalVue()
localVue.use(Vuex)
let wrapper: any

describe('Testing Task Page', () => {
  let actions
  let store
  let getters

  const events = [
    {
        "startTime": 0,
        "endTime": 150
    },
    {
        "startTime": 180,
        "endTime": 300
    },
  ]

  beforeEach(() => {
    actions = {
        fetchEvents: jest.fn(),
    }
    getters = {
        getEvents: () => events,
    }
    store = new Vuex.Store({
        modules: {
            events: {
              namespaced: true,
              state: {
                events: events
              },
              actions,
              getters,
            }
        }

    })

    wrapper = shallowMount(TaskView, {
      localVue,
      store,
    })
  })

  it('Task view is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })

  it('Task Page has proper layout', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Test getter getEvents', () => {
    expect(wrapper.vm.getEvents).toBe(events)
  })

  it('Test eventsBox', () => {
    expect(wrapper.vm.eventsBox).toHaveLength(2)
  })

  it('Check component Row is loaded', () => {
    const row = wrapper.findComponent(Row);
    expect(row.exists()).toBe(true)
  })

  it('Check component EventBox is loaded', () => {
    const eventBox = wrapper.findComponent(EventBox);
    expect(eventBox.exists()).toBe(true)
  })

  it('Test lastItem', () => {
    expect(wrapper.vm.lastItem).toBe(14)
  })

  it('Test allTimeFrames', () => {
    expect(wrapper.vm.allTimeFrames).toHaveLength(15)
  })
})