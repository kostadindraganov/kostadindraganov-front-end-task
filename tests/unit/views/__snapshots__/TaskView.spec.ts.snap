// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Task Page Task Page has proper layout 1`] = `
<main class="container flex flex-col items-center mb-10">
  <h1>Jiminny Calendar</h1>
  <section class="my-">
    <div class="relative mx-3">
      <ul>
        <li>
          <event-box-stub event="[object Object]"></event-box-stub>
        </li>
        <li>
          <event-box-stub event="[object Object]"></event-box-stub>
        </li>
      </ul>
    </div>
    <ul>
      <li>
        <row-stub title="7 AM"></row-stub>
      </li>
      <li>
        <row-stub title="8 AM"></row-stub>
      </li>
      <li>
        <row-stub title="9 AM"></row-stub>
      </li>
      <li>
        <row-stub title="10 AM"></row-stub>
      </li>
      <li>
        <row-stub title="11 AM"></row-stub>
      </li>
      <li>
        <row-stub title="12 PM"></row-stub>
      </li>
      <li>
        <row-stub title="1 PM"></row-stub>
      </li>
      <li>
        <row-stub title="2 PM"></row-stub>
      </li>
      <li>
        <row-stub title="3 PM"></row-stub>
      </li>
      <li>
        <row-stub title="4 PM"></row-stub>
      </li>
      <li>
        <row-stub title="5 PM"></row-stub>
      </li>
      <li>
        <row-stub title="6 PM"></row-stub>
      </li>
      <li>
        <row-stub title="7 PM"></row-stub>
      </li>
      <li>
        <row-stub title="8 PM"></row-stub>
      </li>
      <li>
        <row-stub title="9 PM" islast="true"></row-stub>
      </li>
    </ul>
  </section>
</main>
`;
